cmd/kubeadm/app/cmd/join.go

```diff
@@ -431,7 +431,17 @@ func newJoinData(cmd *cobra.Command, args []string, opt *joinOptions, out io.Wri
 		return nil, err
 	}
 
-	ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(opt.ignorePreflightErrors, cfg.NodeRegistration.IgnorePreflightErrors)
+	initCfg := &kubeadmapi.InitConfiguration{}
+	// If the cfg is not valid config, e.g. this might be a fake config defined in testcase, then lazily fetch `InitConfiguration` until it's really used.
+	if tlsBootstrapCfg == nil {
+		if tlsBootstrapCfg, err = discovery.For(cfg); err == nil {
+			if initCfg, err = fetchInitConfigurationFromJoinConfiguration(cfg, tlsBootstrapCfg); err != nil {
+				return nil, err
+			}
+		}
+	}
+
```

Comment:
hm, wouldn't this result in the init config being fetched always?
how do we know if this is a test config?

also why not add this below:
```
if _, err := os.Stat(adminKubeConfigPath); err == nil && opt.controlPlane {
		// use the admin.conf as tlsBootstrapCfg, that is the kubeconfig file used for reading the kubeadm-config during discovery
		klog.V(1).Infof("[preflight] found %s. Use it for skipping discovery", adminKubeConfigPath)
		tlsBootstrapCfg, err = clientcmd.LoadFromFile(adminKubeConfigPath)
		if err != nil {
			return nil, errors.Wrapf(err, "Error loading %s", adminKubeConfigPath)
		}
	}
```

Reply:
> hm, wouldn't this result in the init config being fetched always?

we cannot get the FG defined from a init config if there is no such config.

> how do we know if this is a test config?

if this is not a test config and an error is hit, then initCfg would be an empty config, then initCfg could be fetched again when it is needed. 

> also why not add this below:

do you mean fetching the config only it's a control plane && admin kubeconfig exist? sorry, I cannot get it, is there any benefit for us to move those code up?

Reply:
> hm, wouldn't this result in the init config being fetched always?
>> we cannot get the FG defined from a init config if there is no such config.

this is pain, is there anyway for us to workaround this?

Reply:
> do you mean fetching the config only it's a control plane && admin kubeconfig exist? sorry, I cannot get it, is there any benefit for us to move those code up?

no just putting the login closer to the block i posted here
https://github.com/kubernetes/kubernetes/pull/119946#discussion_r1315679568

core organization wise

Reply:
i don't think we are understanding each other very well here.
the above logic and comment in the code confused me, but i think you might want to ignore my question for now, because i'm doing N things at once and don't have enough time to have a look at this PR more today.

if you think it's the right thing to do, leave it like so and other reviewers will also have a look.

Reply:
@neolit123 take you time, this is not urgent I think, thanks!

Reply:
> hm, wouldn't this result in the init config being fetched always?

this should be, the reason is we need to get the `initCfg` from cluster, so that we can get the `FeatureGates` from the `initCfg`.

here is the [code](https://github.com/kubernetes/kubernetes/pull/119946/files#diff-c6f6a594b0b45b48a444c274731731e10883c6cf9e2e66f8938aa28af5a57b27R447):
```golang
ignorePreflightErrorsSet, err := validation.ValidateIgnorePreflightErrors(initCfg.FeatureGates, opt.ignorePreflightErrors, cfg.NodeRegistration.IgnorePreflightErrors)
```

Because of that, I add a parameter `lazilyFetchInitCfg` here to keep the original testcase unimpacted, the testcases there expected to lazily fetch a `initConfiguration`, I think that is because testcase should not interact with the cluster. 

>no just putting the login closer to the block i posted here

hmm, looks like I cannot move this piece of code up, because the code must under the `cfg`, I need the `cfg` to get the `cfg.NodeRegistration.IgnorePreflightErrors` as well. 
```golang
cfg, err := configutil.LoadOrDefaultJoinConfiguration(opt.cfgPath, opt.externalcfg)
```